app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb

```diff
@@ -26,40 +31,72 @@ def each_batch
 
   private
 
+  # Get a list of events that have happened since the most recent event time.
+  #
+  # Because Azure event timestamps are not necessarily stamped in order, an
+  # issue occurs where we could accidentally skip over events that happen
+  # in quick succession. We must therefore begin our query a couple minutes
+  # back from our most recent timestamp, and filter out any duplicates.
+  #
+  # See https://bugzilla.redhat.com/show_bug.cgi?id=1724312 for details.
+  #
   def get_events
-    # Grab only events for the last minute if this is the first poll
-    filter = @since ? "eventTimestamp ge #{@since}" : "eventTimestamp ge #{startup_interval}"
-    fields = 'authorization,description,eventName,eventTimestamp,resourceGroupName,resourceProviderName,resourceId,resourceType'
-    events = connection.list(:filter => filter, :select => fields, :all => true).sort_by(&:event_timestamp)
+    filter = "eventTimestamp ge #{most_recent_time}"
+    fields = 'authorization,correlationId,description,eventDataId,eventName,eventTimestamp,operationName,resourceGroupName,resourceProviderName,resourceId,resourceType'
+
+    events = connection.list(:filter => filter, :select => fields, :all => true)
+
+    if events.present?
+      existing_records = EventStream.select(:ems_ref).where(:source => 'AZURE', :ems_ref => events.map(&:event_data_id)).map(&:ems_ref)
```

Comment:
As per a gitter discussion, it turns out we didn't need this at all because `EmsEvent.create_event` is supposed to be filtering out duplicates already. The problem was the timestamp precision from Azure is 7 digits, while Postgres uses 6 digits. The comparison would thus fail, and a duplicate would get added. So, the solution was to truncate the date to 6 digits.